docs/proposals/20230130-go-configuration.md

```diff
@@ -0,0 +1,189 @@
+// Use SidecarConfig struct to attach additonal volumes, service ports and service monitor
+// scrape endpoints for these sidecars.
+func WithSidecars(s k8sutil.SidecarConfig) ObservatoriumAPIOption {
+	return func(a *observatoriumAPI) {
+		a.sidecars = s
+	}
+}
+```
+
+With these options we can define Kubernetes structs for the various Kubernetes Objects needed to deploy Observatorium API, and return them as a map containing the manifest name and [`runtime.Object`](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#Object) (i.e, as `k8sutil.ObjectMap`) via the `Manifests()` method.
+
+```go
+// Manifests creates resultant K8s YAML manifests for customized Observatorium API deployment
+// as a map, containing manifest name and runtime.Object.
+// This provides the ability to override options of the returned manifests
+// in a specific manner (in case the exported ObservatoriumAPIOption functions do not suffice).
+func (c *observatoriumAPI) Manifests(opts ...ObservatoriumAPIOption) k8sutil.ObjectMap {
+```
+
+Finally we can use all these options to simply generate the Observatorium API Kubernetes configuration we need, in `configuration_go/manifests/config/`.
+
+```go
+	// Observatorium API with no sidecar and a serviceMonitor.
+	generator.GenerateWithMimic(
+		g,
+		api.NewObservatoriumAPI(
+			api.WithLogLevel("debug"),
+			api.WithMetrics(
+				"http://observatorium-xyz-thanos-query-frontend.observatorium.svc.cluster.local:9090",
+				"http://observatorium-xyz-thanos-receive.observatorium.svc.cluster.local:19291",
+				"http://observatorium-xyz-rules-objstore.observatorium.svc.cluster.local:8080",
+			),
+			api.WithLogs(
+				"http://observatorium-xyz-loki-query-frontend-http.observatorium.svc.cluster.local:3100",
+				"http://observatorium-xyz-loki-distributor-http.observatorium.svc.cluster.local:3100",
+				"http://observatorium-xyz-loki-ruler-http.observatorium.svc.cluster.local:3100",
+				"http://observatorium-xyz-loki-querier-http.observatorium.svc.cluster.local:3100",
+			),
+			api.WithTraces(
+				"http://observatorium-xyz-jaeger-query.observatorium.svc.cluster.local:16686/",
+				"observatorium-xyz-otel-collector:4317",
+			),
+			api.WithRateLimiter("observatorium-xyz-gubernator.observatorium.svc.cluster.local:8081"),
+			api.WithRBACYAML(string(rbacData)),
+			api.WithTenantsSecret(map[string]string{"tenants.yaml": string(tenantsData)}),
+
+			api.WithImage("quay.io/observatorium/api", "main-2023-01-24-v0.1.2-318-g5f4fdf4"),
+			api.WithName("observatorium-xyz"),
+			api.WithNamespace("observatorium"),
+			api.WithReplicas(3),
+			api.WithAPIResources(apiResources),
+			api.WithServiceMonitor(),
+		).Manifests(),
+		"config",
+	)
+```
+
+### Future Installation Model
+
+We can even leverage [`obsctl`](https://github.com/observatorium/obsctl) in the future to allow generating boilerplate Go code that imports Observatorium Go config and provides the user with commands to generate sane default YAML config.
```

Comment:
Isn't `obsctl` becoming a tool that does many very different things? Is this okay for the project's direction?

Reply:
`obsctl` is a tool for making observatorium easier to interact with IMO. This just generates a boilerplate, and doesn't do anything special, customization is still a manual operation for user. However I don't feel too strongly about this, can remove this idea if others agree. 🙂

Reply:
Totally get the objective of `obsctl`. But it's too generic and going against "do one thing and do it well". Which can be a totally fine direction, so if that's the case, all good to me.

Reply:
I understand your concern @douglascamata, but I wouldn't say it's too much. obsctl still interacts with Observatorium resources in this case.